src/microvmi.rs

```diff
@@ -0,0 +1,207 @@
+    Registers,
+};
+#[cfg(feature = "kvm")]
+use crate::driver::kvm::Kvm;
+#[cfg(feature = "virtualbox")]
+use crate::driver::virtualbox::VBox;
+#[cfg(feature = "xen")]
+use crate::driver::xen::Xen;
+use crate::errors::MicrovmiError;
+use crate::memory::Memory;
+use crate::memory::PaddedMemory;
+use std::cell::RefCell;
+use std::error::Error;
+use std::rc::Rc;
+
+/// Main struct to interact with the library
+pub struct Microvmi {
+    // runtime VMI driver
+    pub(crate) drv: Rc<RefCell<Box<dyn Introspectable>>>,
+    /// Exposes the physical memory as a file-like interface
+    pub memory: Memory,
+    /// Exposes the physical memory as a file-like interface, with padding
+    pub padded_memory: PaddedMemory,
+}
+
+impl Microvmi {
+    /// Initializes a new Microvmi instance
+    ///
+    /// # Arguments
+    ///
+    /// * `domain_name` - The domain name
+    /// * `driver_type` - The driver type to initialize. None will attempt to initialize every driver avaiable
+    /// * `init_option` - Initialization parameters for the driver.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use crate::microvmi::microvmi::Microvmi;
+    /// use crate::microvmi::api::{DriverType, DriverInitParam};
+    /// Microvmi::new("win10", None, None);
+    /// Microvmi::new("win10", Some(DriverType::Xen), None);
+    /// Microvmi::new("win10", Some(DriverType::KVM), Some(DriverInitParam::KVMiSocket("/tmp/introspector".to_string())));
+    /// ```
+    pub fn new(
+        domain_name: &str,
+        driver_type: Option<DriverType>,
+        init_option: Option<DriverInitParam>,
+    ) -> Result<Microvmi, MicrovmiError> {
+        info!("Microvmi init");
+        let drv = match driver_type {
+            None => {
+                // for each possible DriverType
+                let mut driver: Option<Box<dyn Introspectable>> = None;
+                for drv_type in DriverType::into_enum_iter() {
+                    // try to init
+                    match init_driver(domain_name, drv_type, init_option.clone()) {
+                        Ok(drv) => {
+                            driver = Some(drv);
+                            break;
+                        }
+                        Err(e) => {
+                            debug!("{:?} driver initialization failed: {}", drv_type, e);
+                            continue;
+                        }
+                    }
+                }
+                driver.ok_or(MicrovmiError::NoDriverAvailable)?
+            }
+            Some(drv_type) => init_driver(domain_name, drv_type, init_option)?,
+        };
+        let ref_drv = Rc::new(RefCell::new(drv));
+        Ok(Microvmi {
+            drv: ref_drv.clone(),
+            memory: Memory::new(ref_drv.clone())?,
+            padded_memory: PaddedMemory::new(ref_drv.clone())?,
+        })
+    }
+
+    /// Retrieve the number of VCPUs.
+    pub fn get_vcpu_count(&self) -> Result<u16, Box<dyn Error>> {
```

Comment:
You basically implement the `Introspectable` trait without officially implementing the `Introspectable` trait. That's nasty. :/

Reply:
It's also a lot of code duplication and unnecessary indirection. What's so wrong about making the `drv` public in order to allow direct access?